src/java/org/apache/cassandra/index/sai/disk/v1/SegmentMetadataBuilder.java

```diff
@@ -0,0 +1,395 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk.v1;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.marshal.DecimalType;
+import org.apache.cassandra.db.marshal.InetAddressType;
+import org.apache.cassandra.db.marshal.IntegerType;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.disk.PostingList;
+import org.apache.cassandra.index.sai.disk.TermsIterator;
+import org.apache.cassandra.index.sai.disk.format.IndexComponents;
+import org.apache.cassandra.index.sai.disk.v1.kdtree.MutableOneDimPointValues;
+import org.apache.cassandra.index.sai.disk.v6.TermsDistribution;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.cassandra.utils.bytecomparable.ByteSourceInverse;
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * {@link SegmentMetadata} contains a lot of information, so it got its own Builder.
+ * The builder is not only responsible for setting the fields, but also intercepts
+ * the index building process and records the {@link TermsDistribution}.
+ */
+@NotThreadSafe
+public class SegmentMetadataBuilder
+{
+    private static final String HISTOGRAM_SIZE_OPTION = "statistics.histogram_size";
+    private static final String MFT_COUNT_OPTION = "statistics.most_frequent_terms_count";
+
+    private final long segmentRowIdOffset;
+
+    private final AbstractType<?> validator;
+    private final List<Closeable> interceptors = new ArrayList<>();
+
+    private boolean built = false;
+
+    private PrimaryKey minKey;
+    private PrimaryKey maxKey;
+
+    private long minRowId = -1;
+    private long maxRowId = -1;
+
+    private ByteBuffer minTerm;
+    private ByteBuffer maxTerm;
+
+    private long numRows;
```

Comment:
Very minor nit: at the moment, a segment can only have `int` rows.

Reply:
This doesn't cost us much and saves us possible problem if it ever stops being true